TRTLLM-MLA FP8 path

farazkh80 opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

Adds supports for converting MLA query to `fp8_e4m3` and run `flashinfer.decode.trtllm_batch_decode_with_kv_cache_mla` in FP8 path.

## Modifications

- Add query fp8 conversion
- Change server args and docs

## Accuracy Test

<details>
<summary> Flashinfer FP16 deepseek/Deepseek-R1 FP8 model GSM8K and MMLU </summary>

**Server Command**: 
```
python3 -m sglang.launch_server --model-path deepseek-ai/DeepSeek-R1 --trust-remote-code --tp 8 --attention-backend flashinfer --page-size 64
```

**GSM8K Benchmark Command**: 
```
python3 benchmark/gsm8k/bench_sglang.py     --num-shots 5     --num-questions 1319     --parallel 512 --port 30000
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [00:38<00:00, 34.46it/s]
Accuracy: 0.952
Invalid: 0.000
Latency: 38.466 s
Output throughput: 3316.330 token/s
```

**MMLU Benchmark Command**: 
```
python3 benchmark/mmlu/bench_sglang.py --ntrain 5 --data_dir benchmark/mmlu/data --save_dir benchmark/mmlu/results  --parallel 512 --port 30000
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 14042/14042 [01:45<00:00, 132.81it/s]
subject: abstract_algebra, #q:100, acc: 0.750
subject: anatomy, #q:135, acc: 0.844
subject: astronomy, #q:152, acc: 0.941
subject: business_ethics, #q:100, acc: 0.880
subject: clinical_knowledge, #q:265, acc: 0.932
subject: college_biology, #q:144, acc: 0.965
subject: college_chemistry, #q:100, acc: 0.620
subject: college_computer_science, #q:100, acc: 0.840
subject: college_mathematics, #q:100, acc: 0.780
subject: college_medicine, #q:173, acc: 0.873
subject: college_physics, #q:102, acc: 0.833
subject: computer_security, #q:100, acc: 0.900
subject: conceptual_physics, #q:235, acc: 0.928
subject: econometrics, #q:114, acc: 0.763
subject: electrical_engineering, #q:145, acc: 0.869
subject: elementary_mathematics, #q:378, acc: 0.937
subject: formal_logic, #q:126, acc: 0.802
subject: global_facts, #q:100, acc: 0.690
subject: high_school_biology, #q:310, acc: 0.958
subject: high_school_chemistry, #q:203, acc: 0.862
subject: high_school_computer_science, #q:100, acc: 0.940
subject: high_school_european_history, #q:165, acc: 0.885
subject: high_school_geography, #q:198, acc: 0.965
subject: high_school_government_and_politics, #q:193, acc: 0.984
subject: high_school_macroeconomics, #q:390, acc: 0.918
subject: high_school_mathematics, #q:270, acc: 0.741
subject: high_school_microeconomics, #q:238, acc: 0.966
subject: high_school_physics, #q:151, acc: 0.834
subject: high_school_psychology, #q:545, acc: 0.972
subject: high_school_statistics, #q:216, acc: 0.856
subject: high_school_us_history, #q:204, acc: 0.941
subject: high_school_world_history, #q:237, acc: 0.937
subject: human_aging, #q:223, acc: 0.852
subject: human_sexuality, #q:131, acc: 0.924
subject: international_law, #q:121, acc: 0.950
subject: jurisprudence, #q:108, acc: 0.917
subject: logical_fallacies, #q:163, acc: 0.926
subject: machine_learning, #q:112, acc: 0.777
subject: management, #q:103, acc: 0.913
subject: marketing, #q:234, acc: 0.949
subject: medical_genetics, #q:100, acc: 0.950
subject: miscellaneous, #q:783, acc: 0.950
subject: moral_disputes, #q:346, acc: 0.873
subject: moral_scenarios, #q:895, acc: 0.780
subject: nutrition, #q:306, acc: 0.905
subject: philosophy, #q:311, acc: 0.894
subject: prehistory, #q:324, acc: 0.944
subject: professional_accounting, #q:282, acc: 0.865
subject: professional_law, #q:1534, acc: 0.703
subject: professional_medicine, #q:272, acc: 0.956
subject: professional_psychology, #q:612, acc: 0.913
subject: public_relations, #q:110, acc: 0.827
subject: security_studies, #q:245, acc: 0.882
subject: sociology, #q:201, acc: 0.965
subject: us_foreign_policy, #q:100, acc: 0.920
subject: virology, #q:166, acc: 0.590
subject: world_religions, #q:171, acc: 0.936
Total latency: 105.791
Average accuracy: 0.870
```

</details>
<details>
<summary> TRTLLM-MLA FP16 deepseek/Deepseek-R1 FP8 model GSM8K and MMLU </summary>

**Server Command**: 

```
python3 -m sglang.launch_server --model-path deepseek-ai/DeepSeek-R1 --trust-remote-code --tp 8 --attention-backend trtllm_mla
```

**GSM8K Benchmark Command**: 
```
python3 benchmark/gsm8k/bench_sglang.py \
    --num-shots 5 \
    --num-questions 1319 \
    --parallel 512 \
> --port 30000
Downloading from https://raw.githubusercontent.com/openai/grade-school-math/master/grade_school_math/data/test.jsonl to /tmp/test.jsonl
/tmp/test.jsonl: 732kB [00:00, 59.7MB/s]                                                                                                                                                                    
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [01:04<00:00, 20.59it/s]
Accuracy: 0.956
Invalid: 0.000
Latency: 65.608 s
Output throughput: 1921.197 token/s
```

**MMLU Benchmark Command**: 
```
python3 benchmark/mmlu/bench_sglang.py --ntrain 5 --data_dir benchmark/mmlu/data --save_dir benchmark/mmlu/results  --parallel 512 --port 30000
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 14042/14042 [01:47<00:00, 131.03it/s]
subject: abstract_algebra, #q:100, acc: 0.760
subject: anatomy, #q:135, acc: 0.830
subject: astronomy, #q:152, acc: 0.941
subject: business_ethics, #q:100, acc: 0.860
subject: clinical_knowledge, #q:265, acc: 0.928
subject: college_biology, #q:144, acc: 0.972
subject: college_chemistry, #q:100, acc: 0.620
subject: college_computer_science, #q:100, acc: 0.850
subject: college_mathematics, #q:100, acc: 0.760
subject: college_medicine, #q:173, acc: 0.879
subject: college_physics, #q:102, acc: 0.843
subject: computer_security, #q:100, acc: 0.890
subject: conceptual_physics, #q:235, acc: 0.936
subject: econometrics, #q:114, acc: 0.754
subject: electrical_engineering, #q:145, acc: 0.869
subject: elementary_mathematics, #q:378, acc: 0.937
subject: formal_logic, #q:126, acc: 0.817
subject: global_facts, #q:100, acc: 0.670
subject: high_school_biology, #q:310, acc: 0.958
subject: high_school_chemistry, #q:203, acc: 0.882
subject: high_school_computer_science, #q:100, acc: 0.950
subject: high_school_european_history, #q:165, acc: 0.885
subject: high_school_geography, #q:198, acc: 0.960
subject: high_school_government_and_politics, #q:193, acc: 0.984
subject: high_school_macroeconomics, #q:390, acc: 0.926
subject: high_school_mathematics, #q:270, acc: 0.752
subject: high_school_microeconomics, #q:238, acc: 0.958
subject: high_school_physics, #q:151, acc: 0.828
subject: high_school_psychology, #q:545, acc: 0.974
subject: high_school_statistics, #q:216, acc: 0.861
subject: high_school_us_history, #q:204, acc: 0.956
subject: high_school_world_history, #q:237, acc: 0.937
subject: human_aging, #q:223, acc: 0.857
subject: human_sexuality, #q:131, acc: 0.939
subject: international_law, #q:121, acc: 0.950
subject: jurisprudence, #q:108, acc: 0.917
subject: logical_fallacies, #q:163, acc: 0.920
subject: machine_learning, #q:112, acc: 0.795
subject: management, #q:103, acc: 0.922
subject: marketing, #q:234, acc: 0.949
subject: medical_genetics, #q:100, acc: 0.960
subject: miscellaneous, #q:783, acc: 0.957
subject: moral_disputes, #q:346, acc: 0.876
subject: moral_scenarios, #q:895, acc: 0.774
subject: nutrition, #q:306, acc: 0.922
subject: philosophy, #q:311, acc: 0.891
subject: prehistory, #q:324, acc: 0.938
subject: professional_accounting, #q:282, acc: 0.869
subject: professional_law, #q:1534, acc: 0.697
subject: professional_medicine, #q:272, acc: 0.960
subject: professional_psychology, #q:612, acc: 0.915
subject: public_relations, #q:110, acc: 0.827
subject: security_studies, #q:245, acc: 0.882
subject: sociology, #q:201, acc: 0.970
subject: us_foreign_policy, #q:100, acc: 0.920
subject: virology, #q:166, acc: 0.590
subject: world_religions, #q:171, acc: 0.936
Total latency: 107.227
Average accuracy: 0.871
```
</details>
<details>
<summary> TRTLLM-MLA FP8 deepseek/Deepseek-R1 FP8 model GSM8K and MMLU </summary>

```
root@24f704d8370b:/sglang# python3 benchmark/gsm8k/bench_sglang.py     --num-shots 5     --num-questions 1319     --parallel 512 --port 30000
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [01:05<00:00, 20.15it/s]
Accuracy: 0.955
Invalid: 0.000
Latency: 66.912 s
Output throughput: 1920.958 token/s
```

```
root@24f704d8370b:/sglang# python3 benchmark/mmlu/bench_sglang.py --ntrain 5 --data_dir benchmark/mmlu/data --save_dir benchmark/mmlu/results  --parallel 512 --port 30000
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 14042/14042 [02:20<00:00, 99.67it/s]
subject: abstract_algebra, #q:100, acc: 0.750
subject: anatomy, #q:135, acc: 0.844
subject: astronomy, #q:152, acc: 0.941
subject: business_ethics, #q:100, acc: 0.900
subject: clinical_knowledge, #q:265, acc: 0.925
subject: college_biology, #q:144, acc: 0.972
subject: college_chemistry, #q:100, acc: 0.610
subject: college_computer_science, #q:100, acc: 0.840
subject: college_mathematics, #q:100, acc: 0.770
subject: college_medicine, #q:173, acc: 0.879
subject: college_physics, #q:102, acc: 0.833
subject: computer_security, #q:100, acc: 0.890
subject: conceptual_physics, #q:235, acc: 0.928
subject: econometrics, #q:114, acc: 0.754
subject: electrical_engineering, #q:145, acc: 0.869
subject: elementary_mathematics, #q:378, acc: 0.937
subject: formal_logic, #q:126, acc: 0.817
subject: global_facts, #q:100, acc: 0.680
subject: high_school_biology, #q:310, acc: 0.958
subject: high_school_chemistry, #q:203, acc: 0.852
subject: high_school_computer_science, #q:100, acc: 0.950
subject: high_school_european_history, #q:165, acc: 0.879
subject: high_school_geography, #q:198, acc: 0.949
subject: high_school_government_and_politics, #q:193, acc: 0.984
subject: high_school_macroeconomics, #q:390, acc: 0.923
subject: high_school_mathematics, #q:270, acc: 0.748
subject: high_school_microeconomics, #q:238, acc: 0.966
subject: high_school_physics, #q:151, acc: 0.834
subject: high_school_psychology, #q:545, acc: 0.972
subject: high_school_statistics, #q:216, acc: 0.856
subject: high_school_us_history, #q:204, acc: 0.956
subject: high_school_world_history, #q:237, acc: 0.937
subject: human_aging, #q:223, acc: 0.839
subject: human_sexuality, #q:131, acc: 0.939
subject: international_law, #q:121, acc: 0.950
subject: jurisprudence, #q:108, acc: 0.926
subject: logical_fallacies, #q:163, acc: 0.920
subject: machine_learning, #q:112, acc: 0.786
subject: management, #q:103, acc: 0.922
subject: marketing, #q:234, acc: 0.944
subject: medical_genetics, #q:100, acc: 0.950
subject: miscellaneous, #q:783, acc: 0.951
subject: moral_disputes, #q:346, acc: 0.879
subject: moral_scenarios, #q:895, acc: 0.784
subject: nutrition, #q:306, acc: 0.922
subject: philosophy, #q:311, acc: 0.907
subject: prehistory, #q:324, acc: 0.938
subject: professional_accounting, #q:282, acc: 0.869
subject: professional_law, #q:1534, acc: 0.703
subject: professional_medicine, #q:272, acc: 0.956
subject: professional_psychology, #q:612, acc: 0.912
subject: public_relations, #q:110, acc: 0.818
subject: security_studies, #q:245, acc: 0.882
subject: sociology, #q:201, acc: 0.965
subject: us_foreign_policy, #q:100, acc: 0.930
subject: virology, #q:166, acc: 0.596
subject: world_religions, #q:171, acc: 0.930
Total latency: 140.947
Average accuracy: 0.871
```


</details>
<!-- If this PR affects model-side code (e.g., kernels, model architecture), please provide accuracy test results. Ref: https://docs.sglang.ai/references/accuracy_evaluation.html -->

## Benchmark & Profiling

<details>
  <summary>Benchmarking random range ratio 0.0</summary>

## 1000→1000 Tokens

| Backend | Precision | Batch Size | Total Requests | Max Concurrency | Achieved Concurrency | Output Throughput (tok/s) | Perf Gain Output vs Flashinfer FP16 | Perf Gain Output vs TrtLLM FP16 | Total Throughput (tok/s) | Perf Gain Total vs Flashinfer FP16 | Perf Gain Total vs TrtLLM FP16 | Total Input Tokens | Total Output Tokens |
|---------|-----------|------------|----------------|-----------------|----------------------|---------------------------|-------------------------------------|--------------------------------|--------------------------|------------------------------------|---------------------------------|--------------------------|--------------------------|
| flashinfer | fp16 | 1 | 1 | 1 | 1.0 | 76 | N/A | N/A | 88 | N/A | N/A | 38 | 235 |
| trtllm_mla | fp16 | 1 | 1 | 1 | 1.0 | 76 | 1.00 | N/A | 89 | 1.00 | N/A | 38 | 235 |
| trtllm_mla | fp8 | 1 | 1 | 1 | 1.0 | 76 | 0.99 | 0.99 | 88 | 0.99 | 0.99 | 38 | 235 |
| flashinfer | fp16 | 64 | 64 | 64 | 37.0 | 1205 | N/A | N/A | 2449 | N/A | N/A | 32529 | 31524 |
| trtllm_mla | fp16 | 64 | 64 | 64 | 38.4 | 1198 | 0.99 | N/A | 2434 | 0.99 | N/A | 32529 | 31524 |
| trtllm_mla | fp8 | 64 | 64 | 64 | 38.2 | 1210 | 1.00 | 1.01 | 2460 | 1.00 | 1.01 | 32529 | 31524 |
| flashinfer | fp16 | 128 | 128 | 128 | 76.6 | 2081 | N/A | N/A | 4131 | N/A | N/A | 64117 | 65066 |
| trtllm_mla | fp16 | 128 | 128 | 128 | 77.6 | 2107 | 1.01 | N/A | 4184 | 1.01 | N/A | 64117 | 65066 |
| trtllm_mla | fp8 | 128 | 128 | 128 | 76.2 | 2191 | 1.05 | 1.04 | 4350 | 1.05 | 1.04 | 64117 | 65066 |
| flashinfer | fp16 | 256 | 256 | 256 | 157.8 | 3119 | N/A | N/A | 6316 | N/A | N/A | 129311 | 126170 |
| trtllm_mla | fp16 | 256 | 256 | 256 | 158.6 | 3310 | 1.06 | N/A | 6703 | 1.06 | N/A | 129311 | 126170 |
| trtllm_mla | fp8 | 256 | 256 | 256 | 159.7 | 3360 | 1.08 | 1.02 | 6804 | 1.08 | 1.02 | 129311 | 126170 |
| flashinfer | fp16 | 512 | 512 | 512 | 326.9 | 4411 | N/A | N/A | 8824 | N/A | N/A | 255737 | 255587 |
| trtllm_mla | fp16 | 512 | 512 | 512 | 335.8 | 4696 | 1.06 | N/A | 9395 | 1.06 | N/A | 255737 | 255587 |
| trtllm_mla | fp8 | 512 | 512 | 512 | 337.8 | 4846 | 1.10 | 1.03 | 9696 | 1.10 | 1.03 | 255737 | 255587 |


## 8192→1000 Tokens

| Backend | Precision | Batch Size | Total Requests | Max Concurrency | Achieved Concurrency | Output Throughput (tok/s) | Perf Gain Output vs Flashinfer FP16 | Perf Gain Output vs TrtLLM FP16 | Total Throughput (tok/s) | Perf Gain Total vs Flashinfer FP16 | Perf Gain Total vs TrtLLM FP16 | Total Input Tokens | Total Output Tokens |
|---------|-----------|------------|----------------|-----------------|----------------------|---------------------------|-------------------------------------|--------------------------------|--------------------------|------------------------------------|---------------------------------|--------------------------|--------------------------|
| flashinfer | fp16 | 1 | 1 | 1 | 1.0 | 70 | N/A | N/A | 1603 | N/A | N/A | 5158 | 235 |
| trtllm_mla | fp16 | 1 | 1 | 1 | 1.0 | 65 | 0.93 | N/A | 1488 | 0.93 | N/A | 5158 | 235 |
| trtllm_mla | fp8 | 1 | 1 | 1 | 1.0 | 62 | 0.89 | 0.96 | 1434 | 0.89 | 0.96 | 5158 | 235 |
| flashinfer | fp16 | 64 | 64 | 64 | 44.1 | 807 | N/A | N/A | 7480 | N/A | N/A | 263118 | 31827 |
| trtllm_mla | fp16 | 64 | 64 | 64 | 44.1 | 885 | 1.10 | N/A | 8203 | 1.10 | N/A | 263118 | 31827 |
| trtllm_mla | fp8 | 64 | 64 | 64 | 44.2 | 912 | 1.13 | 1.03 | 8455 | 1.13 | 1.03 | 263118 | 31827 |
| flashinfer | fp16 | 128 | 128 | 128 | 92.8 | 1108 | N/A | N/A | 9728 | N/A | N/A | 502881 | 64662 |
| trtllm_mla | fp16 | 128 | 128 | 128 | 94.0 | 1240 | 1.12 | N/A | 10884 | 1.12 | N/A | 502881 | 64662 |
| trtllm_mla | fp8 | 128 | 128 | 128 | 94.6 | 1323 | 1.19 | 1.07 | 11612 | 1.19 | 1.07 | 502881 | 64662 |
| flashinfer | fp16 | 256 | 256 | 256 | 195.7 | 1277 | N/A | N/A | 11761 | N/A | N/A | 1040965 | 126805 |
| trtllm_mla | fp16 | 256 | 256 | 256 | 200.8 | 1521 | 1.19 | N/A | 14005 | 1.19 | N/A | 1040965 | 126805 |
| trtllm_mla | fp8 | 256 | 256 | 256 | 204.0 | 1648 | 1.29 | 1.08 | 15179 | 1.29 | 1.08 | 1040965 | 126805 |
| flashinfer | fp16 | 512 | 512 | 512 | 363.7 | 1304 | N/A | N/A | 12136 | N/A | N/A | 2121577 | 255499 |
| trtllm_mla | fp16 | 512 | 512 | 512 | 370.8 | 1618 | 1.24 | N/A | 15052 | 1.24 | N/A | 2121577 | 255499 |
| trtllm_mla | fp8 | 512 | 512 | 512 | 431.5 | 1938 | 1.49 | 1.20 | 18034 | 1.49 | 1.20 | 2121577 | 255499 |


## 1000→8192 Tokens

| Backend | Precision | Batch Size | Total Requests | Max Concurrency | Achieved Concurrency | Output Throughput (tok/s) | Perf Gain Output vs Flashinfer FP16 | Perf Gain Output vs TrtLLM FP16 | Total Throughput (tok/s) | Perf Gain Total vs Flashinfer FP16 | Perf Gain Total vs TrtLLM FP16 | Total Input Tokens | Total Output Tokens |
|---------|-----------|------------|----------------|-----------------|----------------------|---------------------------|-------------------------------------|--------------------------------|--------------------------|------------------------------------|---------------------------------|--------------------------|--------------------------|
| flashinfer | fp16 | 1 | 1 | 1 | 1.0 | 76 | N/A | N/A | 88 | N/A | N/A | 38 | 235 |
| trtllm_mla | fp16 | 1 | 1 | 1 | 1.0 | 76 | 1.00 | N/A | 89 | 1.00 | N/A | 38 | 235 |
| trtllm_mla | fp8 | 1 | 1 | 1 | 1.0 | 75 | 0.99 | 0.99 | 88 | 0.99 | 0.99 | 38 | 235 |
| flashinfer | fp16 | 64 | 64 | 64 | 31.9 | 1115 | N/A | N/A | 1270 | N/A | N/A | 32529 | 234771 |
| trtllm_mla | fp16 | 64 | 64 | 64 | 32.3 | 1218 | 1.09 | N/A | 1386 | 1.09 | N/A | 32529 | 234771 |
| trtllm_mla | fp8 | 64 | 64 | 64 | 32.4 | 1258 | 1.13 | 1.03 | 1433 | 1.13 | 1.03 | 32529 | 234771 |
| flashinfer | fp16 | 128 | 128 | 128 | 70.1 | 1769 | N/A | N/A | 1986 | N/A | N/A | 64117 | 522611 |
| trtllm_mla | fp16 | 128 | 128 | 128 | 70.8 | 2158 | 1.22 | N/A | 2423 | 1.22 | N/A | 64117 | 522611 |
| trtllm_mla | fp8 | 128 | 128 | 128 | 70.7 | 2285 | 1.29 | 1.06 | 2565 | 1.29 | 1.06 | 64117 | 522611 |
| flashinfer | fp16 | 256 | 256 | 256 | 146.9 | 2532 | N/A | N/A | 2829 | N/A | N/A | 129311 | 1103959 |
| trtllm_mla | fp16 | 256 | 256 | 256 | 150.0 | 3457 | 1.37 | N/A | 3862 | 1.37 | N/A | 129311 | 1103959 |
| trtllm_mla | fp8 | 256 | 256 | 256 | 150.5 | 3840 | 1.52 | 1.11 | 4290 | 1.52 | 1.11 | 129311 | 1103959 |
| flashinfer | fp16 | 512 | 512 | 512 | 279.5 | 3309 | N/A | N/A | 3707 | N/A | N/A | 255737 | 2126946 |
| trtllm_mla | fp16 | 512 | 512 | 512 | 289.5 | 4729 | 1.43 | N/A | 5297 | 1.43 | N/A | 255737 | 2126946 |
| trtllm_mla | fp8 | 512 | 512 | 512 | 302.0 | 5574 | 1.68 | 1.18 | 6244 | 1.68 | 1.18 | 255737 | 2126946 |


</details>

<details>
  <summary>Benchmarking random range ratio 1.0</summary>

## 1024→1024 Tokens

| Backend | Precision | Batch Size | Total Requests | Max Concurrency | Achieved Concurrency | Output Throughput (tok/s) | Perf Gain Output vs Flashinfer FP16 | Perf Gain Output vs TrtLLM FP16 | Total Throughput (tok/s) | Perf Gain Total vs Flashinfer FP16 | Perf Gain Total vs TrtLLM FP16 | Total Input Tokens | Total Output Tokens |
|---------|-----------|------------|----------------|-----------------|----------------------|---------------------------|-------------------------------------|--------------------------------|--------------------------|------------------------------------|---------------------------------|--------------------------|--------------------------|
| flashinfer | fp16 | 1 | 1 | 1 | 1.0 | 80 | N/A | N/A | 160 | N/A | N/A | 1024 | 1024 |
| trtllm_mla | fp16 | 1 | 1 | 1 | 1.0 | 79 | 0.99 | N/A | 159 | 0.99 | N/A | 1024 | 1024 |
| trtllm_mla | fp8 | 1 | 1 | 1 | 1.0 | 78 | 0.97 | 0.98 | 155 | 0.97 | 0.98 | 1024 | 1024 |
| flashinfer | fp16 | 64 | 64 | 64 | 64.0 | 1864 | N/A | N/A | 3729 | N/A | N/A | 65536 | 65536 |
| trtllm_mla | fp16 | 64 | 64 | 64 | 64.0 | 1969 | 1.06 | N/A | 3937 | 1.06 | N/A | 65536 | 65536 |
| trtllm_mla | fp8 | 64 | 64 | 64 | 64.0 | 1998 | 1.07 | 1.01 | 3995 | 1.07 | 1.01 | 65536 | 65536 |
| flashinfer | fp16 | 128 | 128 | 128 | 127.9 | 2959 | N/A | N/A | 5917 | N/A | N/A | 131072 | 131072 |
| trtllm_mla | fp16 | 128 | 128 | 128 | 127.9 | 3314 | 1.12 | N/A | 6627 | 1.12 | N/A | 131072 | 131072 |
| trtllm_mla | fp8 | 128 | 128 | 128 | 127.9 | 3427 | 1.16 | 1.03 | 6854 | 1.16 | 1.03 | 131072 | 131072 |
| flashinfer | fp16 | 256 | 256 | 256 | 255.8 | 4184 | N/A | N/A | 8369 | N/A | N/A | 262144 | 262144 |
| trtllm_mla | fp16 | 256 | 256 | 256 | 255.8 | 4823 | 1.15 | N/A | 9646 | 1.15 | N/A | 262144 | 262144 |
| trtllm_mla | fp8 | 256 | 256 | 256 | 255.8 | 5018 | 1.20 | 1.04 | 10036 | 1.20 | 1.04 | 262144 | 262144 |
| flashinfer | fp16 | 512 | 512 | 512 | 511.6 | 5075 | N/A | N/A | 10151 | N/A | N/A | 524288 | 524288 |
| trtllm_mla | fp16 | 512 | 512 | 512 | 511.5 | 6077 | 1.20 | N/A | 12154 | 1.20 | N/A | 524288 | 524288 |
| trtllm_mla | fp8 | 512 | 512 | 512 | 511.5 | 6414 | 1.26 | 1.06 | 12827 | 1.26 | 1.06 | 524288 | 524288 |

# 8192→1024 Tokens

| Backend | Precision | Batch Size | Total Requests | Max Concurrency | Achieved Concurrency | Output Throughput (tok/s) | Perf Gain Output vs Flashinfer FP16 | Perf Gain Output vs TrtLLM FP16 | Total Throughput (tok/s) | Perf Gain Total vs Flashinfer FP16 | Perf Gain Total vs TrtLLM FP16 | Total Input Tokens | Total Output Tokens |
|---------|-----------|------------|----------------|-----------------|----------------------|---------------------------|-------------------------------------|--------------------------------|--------------------------|------------------------------------|---------------------------------|--------------------------|--------------------------|
| flashinfer | fp16 | 1 | 1 | 1 | 1.0 | 74 | N/A | N/A | 663 | N/A | N/A | 8192 | 1024 |
| trtllm_mla | fp16 | 1 | 1 | 1 | 1.0 | 75 | 1.02 | N/A | 676 | 1.02 | N/A | 8192 | 1024 |
| trtllm_mla | fp8 | 1 | 1 | 1 | 1.0 | 76 | 1.03 | 1.01 | 682 | 1.03 | 1.01 | 8192 | 1024 |
| flashinfer | fp16 | 64 | 64 | 64 | 64.0 | 884 | N/A | N/A | 7953 | N/A | N/A | 524288 | 65536 |
| trtllm_mla | fp16 | 64 | 64 | 64 | 64.0 | 1248 | 1.41 | N/A | 11232 | 1.41 | N/A | 524288 | 65536 |
| trtllm_mla | fp8 | 64 | 64 | 64 | 64.0 | 1341 | 1.52 | 1.07 | 12065 | 1.52 | 1.07 | 524288 | 65536 |
| flashinfer | fp16 | 128 | 128 | 128 | 128.0 | 1069 | N/A | N/A | 9625 | N/A | N/A | 1048576 | 131072 |
| trtllm_mla | fp16 | 128 | 128 | 128 | 128.0 | 1501 | 1.40 | N/A | 13513 | 1.40 | N/A | 1048576 | 131072 |
| trtllm_mla | fp8 | 128 | 128 | 128 | 128.0 | 1652 | 1.55 | 1.10 | 14870 | 1.55 | 1.10 | 1048576 | 131072 |
| flashinfer | fp16 | 256 | 256 | 256 | 191.0 | 1043 | N/A | N/A | 9387 | N/A | N/A | 2097152 | 262144 |
| trtllm_mla | fp16 | 256 | 256 | 256 | 192.0 | 1475 | 1.41 | N/A | 13274 | 1.41 | N/A | 2097152 | 262144 |
| trtllm_mla | fp8 | 256 | 256 | 256 | 255.9 | 1894 | 1.82 | 1.28 | 17046 | 1.82 | 1.28 | 2097152 | 262144 |
| flashinfer | fp16 | 512 | 512 | 512 | 318.6 | 1042 | N/A | N/A | 9375 | N/A | N/A | 4194304 | 524288 |
| trtllm_mla | fp16 | 512 | 512 | 512 | 320.4 | 1478 | 1.42 | N/A | 13298 | 1.42 | N/A | 4194304 | 524288 |
| trtllm_mla | fp8 | 512 | N/A | N/A | N/A | N/A | N/A | N/A | N/A | N/A | N/A | 4194304 | 524288 |

# 1024→8192 Tokens

| Backend | Precision | Batch Size | Total Requests | Max Concurrency | Achieved Concurrency | Output Throughput (tok/s) | Perf Gain Output vs Flashinfer FP16 | Perf Gain Output vs TrtLLM FP16 | Total Throughput (tok/s) | Perf Gain Total vs Flashinfer FP16 | Perf Gain Total vs TrtLLM FP16 | Total Input Tokens | Total Output Tokens |
|---------|-----------|------------|----------------|-----------------|----------------------|---------------------------|-------------------------------------|--------------------------------|--------------------------|------------------------------------|---------------------------------|--------------------------|--------------------------|
| flashinfer | fp16 | 1 | 1 | 1 | 1.0 | 77 | N/A | N/A | 86 | N/A | N/A | 1024 | 8192 |
| trtllm_mla | fp16 | 1 | 1 | 1 | 1.0 | 79 | 1.03 | N/A | 89 | 1.03 | N/A | 1024 | 8192 |
| trtllm_mla | fp8 | 1 | 1 | 1 | 1.0 | 80 | 1.04 | 1.01 | 90 | 1.04 | 1.01 | 1024 | 8192 |
| flashinfer | fp16 | 64 | 64 | 64 | 64.0 | 1531 | N/A | N/A | 1723 | N/A | N/A | 65536 | 524288 |
| trtllm_mla | fp16 | 64 | 64 | 64 | 64.0 | 2067 | 1.35 | N/A | 2326 | 1.35 | N/A | 65536 | 524288 |
| trtllm_mla | fp8 | 64 | 64 | 64 | 64.0 | 2239 | 1.46 | 1.08 | 2519 | 1.46 | 1.08 | 65536 | 524288 |
| flashinfer | fp16 | 128 | 128 | 128 | 128.0 | 2181 | N/A | N/A | 2454 | N/A | N/A | 131072 | 1048576 |
| trtllm_mla | fp16 | 128 | 128 | 128 | 128.0 | 3282 | 1.50 | N/A | 3692 | 1.50 | N/A | 131072 | 1048576 |
| trtllm_mla | fp8 | 128 | 128 | 128 | 128.0 | 3743 | 1.72 | 1.14 | 4211 | 1.72 | 1.14 | 131072 | 1048576 |
| flashinfer | fp16 | 256 | N/A | N/A | N/A | N/A | N/A | N/A | N/A | N/A | N/A | 262144 | 2097152 |
| trtllm_mla | fp16 | 256 | N/A | N/A | N/A | N/A | N/A | N/A | N/A | N/A | N/A | 262144 | 2097152 |
| trtllm_mla | fp8 | 256 | 256 | 256 | 256.0 | 5498 | N/A | N/A | 6185 | N/A | N/A | 262144 | 2097152 |
| flashinfer | fp16 | 512 | N/A | N/A | N/A | N/A | N/A | N/A | N/A | N/A | N/A | 524288 | 4194304 |
| trtllm_mla | fp16 | 512 | N/A | N/A | N/A | N/A | N/A | N/A | N/A | N/A | N/A | 524288 | 4194304 |
| trtllm_mla | fp8 | 512 | N/A | N/A | N/A | N/A | N/A | N/A | N/A | N/A | N/A | 524288 | 4194304 |
</details>



<!-- If this PR is expected to impact performance, please provide benchmark and profiling results. Ref: https://docs.sglang.ai/references/benchmark_and_profiling.html -->

## Checklist

- [X] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [X] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [X] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [X] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

farazkh80 committed:
fp8 works

Signed-off-by: Faraz Khoubsirat <58580514+farazkh80@users.noreply.github.com>

farazkh80 committed:
added docs and test

Signed-off-by: Faraz Khoubsirat <58580514+farazkh80@users.noreply.github.com>

merrymercy commented on python/pyproject.toml:
```diff
@@ -75,6 +75,7 @@ blackwell = [
     "tiktoken",
 ]
 
+
```
revert this

merrymercy commented:
Could you also fix the conflict. Thanks!

farazkh80 committed:
remove todo

Signed-off-by: Faraz Khoubsirat <58580514+farazkh80@users.noreply.github.com>

farazkh80 committed:
add rope for fp8

Signed-off-by: Faraz Khoubsirat <58580514+farazkh80@users.noreply.github.com>

farazkh80 committed:
undo change

Signed-off-by: Faraz Khoubsirat <58580514+farazkh80@users.noreply.github.com>

farazkh80 commented:
need to fix test and profile runtime to see if it is indeed faster than concat or no

fzyzcjy commented:
looking forward to this; btw, I would appreciate it if know whether we have PR for prefill fp8?

elfiegg commented:
@fzyzcjy wondering if you have perf analysis indicating prefill could be benefited from FP8

pavanimajety commented:
Pre-requisite: https://github.com/flashinfer-ai/flashinfer/pull/1339

farazkh80 committed:
test still failing but some interface changes

Signed-off-by: Faraz Khoubsirat <58580514+farazkh80@users.noreply.github.com>

fzyzcjy commented:
@elfiegg No, but from my naive understanding, fp8 computation may be faster than bf16 since we have more theoretical flops, and it would be great to fuse quant into e.g. rope because whenever we use fp8 for decode our prefill output kv cache also needs to be in fp8

pavanimajety commented on python/sglang/srt/layers/attention/trtllm_mla_backend.py:
```diff
@@ -287,38 +287,115 @@ def init_forward_metadata(self, forward_batch: ForwardBatch):
+        quantizes all components to FP8, and merges the query components into a single tensor.
+        
+        Args:
+            q_nope: Query no-position-encoding component [seq_len, num_heads, kv_lora_rank]
```
small suggestion: add the expected dtype for each of query and k inputs and the expected outputs for clarity

farazkh80 commented on python/sglang/test/attention/test_trtllm_mla_backend.py:
```diff
@@ -63,19 +63,37 @@
+            "description": "Single FP16 vs reference",
         },
+        # {
+        #     "name": "single_fp8",
```
they are currently failing when comparing accuracy against flashinfer MLA that happens in BF16, I need to fix them.

farazkh80 replied:
fixed now

elfiegg commented on python/sglang/srt/layers/attention/trtllm_mla_backend.py:
```diff
@@ -287,38 +287,115 @@ def init_forward_metadata(self, forward_batch: ForwardBatch):
+        if self.data_type == torch.float8_e4m3fn:
+            # For FP8 path, we quantize the query and rope parts and merge them into a single tensor
+            # Note: rope application in deepseek_v2.py:forward_absorb_prepare is skipped for FP8 decode path of this trtllm_mla backend
+            assert all(x is not None for x in [q_rope, k_rope, cos_sin_cache])
```
do me a favor to throw something meaningful after this assertion? e.g. what's wrong / alternative solution?

farazkh80 committed:
test passing and lint

Signed-off-by: Faraz Khoubsirat <58580514+farazkh80@users.noreply.github.com>

farazkh80 committed:
pr comment add dtype

Signed-off-by: Faraz Khoubsirat <58580514+farazkh80@users.noreply.github.com>

farazkh80 commented:
This PR is ready for review @fzyzcjy @merrymercy 

Ran latest benchmarking and hightlights include
- With TRTLLM MLA fp8 We can get up to 70-80% throughput gain compared to flashinfer MLA. i.e. on ISL:1024-OSL:8192

Backend | Precision  | Total Requests | Max Concurrency | Output Throughput (tok/s) |  Total Throughput (tok/s) |
-- | -- | -- | -- | -- | -- |
flashinfer | fp16  | 128 | 128 | 2106  | 2369 | 
trtllm_mla | fp16 |  128 | 128 | 315  | 3551  |
trtllm_mla | fp8 | 128 | 128 | 3685 |  4145 |

- On ISL:1024-OSL:1024 with trtllm mla fp8 path we can achieve concurrency 1024 which OOMs with fp16 path (both trtllm-mla and flashinfer backends) 
- On ISL:1024-OSL:8192 with trtllm mla fp8 path we can achieve concurrency 256 which OOMs with fp16 path (both trtllm-mla and flashinfer backends) 


In terms of accuracy here are the measures

```
TRTLLM FP16
root@06df55f8f415:/sglang# python3 benchmark/gsm8k/bench_sglang.py     --num-shots 5     --num-questions 1319     --parallel 512 --port 30000
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [00:37<00:00, 34.86it/s]
Accuracy: 0.957
Invalid: 0.000


Flashinfer MLA fp16
python3 benchmark/gsm8k/bench_sglang.py     --num-shots 5     --num-questions 1319     --parallel 512 --port 30000
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [00:38<00:00, 34.46it/s]
Accuracy: 0.952
Invalid: 0.000


TRTLLM fp8
root@06df55f8f415:/sglang# python3 benchmark/gsm8k/bench_sglang.py     --num-shots 5     --num-questions 1319     --parallel 512 --port 30000
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [01:13<00:00, 17.84it/s]
Accuracy: 0.956
Invalid: 0.000
```

fzyzcjy reviewed (commented):
LGTM, some nits

farazkh80 commented on python/sglang/srt/layers/attention/trtllm_mla_backend.py:
```diff
@@ -287,38 +287,135 @@ def init_forward_metadata(self, forward_batch: ForwardBatch):
             q_rope_reshaped = q_rope.view(
                 -1, layer.tp_q_head_num, layer.head_dim - layer.v_head_dim
             )
             query = torch.cat([q_nope, q_rope_reshaped], dim=-1)
```
yes only used when in fp16. But since fp8 is faster and no accuracy degradation, fp8 should be default if possible

fzyzcjy replied:
Yes I think so

farazkh80 committed:
pr comments

Signed-off-by: Faraz Khoubsirat <58580514+farazkh80@users.noreply.github.com>

farazkh80 commented on python/sglang/srt/layers/attention/trtllm_mla_backend.py:
```diff
@@ -287,38 +287,135 @@ def init_forward_metadata(self, forward_batch: ForwardBatch):
+            assert (
+                k is not None and k_rope is not None
+            ), "For populating trtllm_mla both k_nope and k_rope should be not None."
+            forward_batch.token_to_kv_pool.set_mla_kv_buffer(
```
done

farazkh80 replied:
I actually remove the assert since I realized in `forward_absorb_core` in `deepseek_v2.py` I use the same attention call where v is not passed as None. @fzyzcjy 

```python
if (
    self.current_attention_backend == "fa3"
    or self.current_attention_backend == "flashinfer"
    or self.current_attention_backend == "cutlass_mla"
    or self.current_attention_backend == "trtllm_mla"
):
        extra_args = {}
        if self._enable_fuse_rope(forward_batch):
            extra_args = {
                "cos_sin_cache": self.rotary_emb.cos_sin_cache,
                "is_neox": self.rotary_emb.is_neox_style,
            }
        attn_output = self.attn_mqa(
            q_nope_out,
            k_nope,
            k_nope,
            forward_batch,
            q_rope=q_pe,
            k_rope=k_pe,
            **extra_args,
        )
```

farazkh80 commented on python/sglang/srt/server_args.py:
```diff
@@ -435,16 +435,21 @@ def print_deprecated_warning(message: str):
+            if self.speculative_algorithm is not None:
+                raise ValueError(
+                    "trtllm_mla backend does not support speculative decoding yet."
+                )
```
done

farazkh80 commented on python/sglang/srt/models/deepseek_v2.py:
```diff
@@ -1273,7 +1273,13 @@ def forward_absorb_prepare(
+            self.current_attention_backend == "trtllm_mla"
+            and forward_batch.forward_mode.is_decode_or_idle()
+            and forward_batch.attn_backend.data_type == torch.float8_e4m3fn
+        ):
```
done

farazkh80 commented:
addressed all comments @fzyzcjy please let me know if it all looks good.

In terms of merging @zhyncs , we still need flashinfer 0.2.11 to be released and updated on upstream (or I can modify in this PR) since the rope fusion is a part of [v0.2.11](https://github.com/flashinfer-ai/flashinfer/releases/tag/v0.2.11)

farazkh80 committed:
update name

Signed-off-by: Faraz Khoubsirat <58580514+farazkh80@users.noreply.github.com>

farazkh80 committed:
revert one change

Signed-off-by: Faraz Khoubsirat <58580514+farazkh80@users.noreply.github.com>

fzyzcjy commented on python/sglang/srt/models/deepseek_v2.py:
```diff
@@ -1194,6 +1194,13 @@ def forward_normal_core(self, q, k, v, forward_batch):
         output, _ = self.o_proj(attn_output)
         return output
 
+    def _enable_fuse_rope(self, forward_batch: ForwardBatch) -> bool:
```
nit: maybe we can name it to fuse_rope_to_blahblah or sth like that, b/c today I also write some code about "fuse rope with set-kv-buffer" and it seems different from that

fzyzcjy commented on python/sglang/srt/server_args.py:
```diff
@@ -432,17 +432,26 @@ def print_deprecated_warning(message: str):
                     "TRTLLM MLA backend is only supported on Blackwell GPUs (SM100). Please use a different backend."
                 )
 
+            if self.kv_cache_dtype not in ["fp8_e4m3", "fp16", "auto"]:
```
nit: is it bf16 or fp16

farazkh80 replied:
I removed the fp16/bf16 since it is not supported in the server_args for kv-cache-dtype anyways. Valid options are `fp8_e4m3` for fp8 path and `auto` for BFloat16 path for trtllm-mla.

```
fmhaSm100Kernel_QkvBfloat16OBfloat16HQk576HV512HVPerCta512PagedKvDenseP64VarSeqQ16Kv128PersistentSwapsAbForGen
```
FP8 kernel cubin 
```
fmhaSm100Kernel_QkvE4m3OBfloat16HQk576HV512HVPerCta512PagedKvDenseP64VarSeqQ16Kv128PersistentSwapsAbForGen
```

fzyzcjy reviewed (approved):
LGTM, only two nits

fzyzcjy commented:
> addressed all comments @fzyzcjy please let me know if it all looks good.

Yes it LGTM (with minor nits)

> we still need flashinfer 0.2.11 to be released 

that is on main now

and btw could you please resolve merge conflict (looks like .md has conflict which may be easy to solve)

farazkh80 committed:
pr comments nit picks

Signed-off-by: Faraz Khoubsirat <58580514+farazkh80@users.noreply.github.com>

farazkh80 commented:
@fzyzcjy I had to do a rebase+force push to solve the conflicts. Addressed all the comments now. Let me know if it looks good to run CI and merge

farazkh80 committed:
lint

Signed-off-by: Faraz Khoubsirat <58580514+farazkh80@users.noreply.github.com>
